src/compiler/checker.ts

```diff
@@ -36458,9 +36534,50 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
             allTypesAssignableToKind(leftType, TypeFlags.Primitive)) {
             error(left, Diagnostics.The_left_hand_side_of_an_instanceof_expression_must_be_of_type_any_an_object_type_or_a_type_parameter);
         }
-        // NOTE: do not raise error if right is unknown as related error was already reported
-        if (!(isTypeAny(rightType) || typeHasCallOrConstructSignatures(rightType) || isTypeSubtypeOf(rightType, globalFunctionType))) {
-            error(right, Diagnostics.The_right_hand_side_of_an_instanceof_expression_must_be_of_type_any_or_of_a_type_assignable_to_the_Function_interface_type);
+        if (!isTypeAny(rightType)) {
+            // if rightType is an object type with a custom `[Symbol.hasInstance]` method, then it is potentially
+            // valid on the right-hand side of the `instanceof` operator. This allows normal `object` types to
+            // participate in `instanceof`, as per Step 2 of https://tc39.es/ecma262/#sec-instanceofoperator.
+            const hasInstanceMethodType = getSymbolHasInstanceMethodOfObjectType(rightType);
+            if (hasInstanceMethodType) {
+                // avoid a complex check for every `instanceof` when the `[Symbol.hasInstance]` method has a single
+                // call signature that neither restricts nor narrows (via type predicate) the LHS value, e.g.
+                // `(value: unknown) => boolean`.
+                const cache = hasInstanceMethodType as HasInstanceMethodType;

```

Comment:
I'm curious about how you know this optimization is worth it.

Reply:
At the moment, purely based on experience, though in general I would validate this assumption through benchmarking. Without a cache we would be performing a number of additional checks over and over again against the default implementation of `[Symbol.hasInstance](value)` on `Function` for every occurrence of `instanceof` in a given codebase, and I'd like to avoid performing additional work against existing code.

This does incur additional polymorphism in `Type`, but `Type` is already megamorphic and I'd prefer to work on decreasing `Type` polymorphism in a separate PR.